ast/src/parsed/mod.rs

```diff
@@ -52,6 +55,89 @@ pub enum PilStatement<T> {
     Expression(usize, Expression<T>),
 }
 
+impl<T> PilStatement<T> {
+    /// If the statement is a symbol definition, returns all (local) names of defined symbols.
+    pub fn symbol_definition_names(&self) -> Box<dyn Iterator<Item = &String> + '_> {
+        match self {
+            PilStatement::PolynomialDefinition(_, name, _)
+            | PilStatement::PolynomialConstantDefinition(_, name, _)
+            | PilStatement::ConstantDefinition(_, name, _)
+            | PilStatement::PublicDeclaration(_, name, _, _, _)
+            | PilStatement::LetStatement(_, name, _) => Box::new(once(name)),
+            PilStatement::PolynomialConstantDeclaration(_, polynomials)
+            | PilStatement::PolynomialCommitDeclaration(_, polynomials, _) => {
+                Box::new(polynomials.iter().map(|p| &p.name))
+            }
+
+            PilStatement::Include(_, _)
+            | PilStatement::Namespace(_, _, _)
+            | PilStatement::PolynomialIdentity(_, _)
+            | PilStatement::PlookupIdentity(_, _, _)
+            | PilStatement::PermutationIdentity(_, _, _)
+            | PilStatement::ConnectIdentity(_, _, _)
+            | PilStatement::Expression(_, _) => Box::new(empty()),
+        }
+    }
+
+    /// Returns an iterator over all (top-level) expressions in this statement.
+    pub fn expressions(&self) -> Box<dyn Iterator<Item = &Expression<T>> + '_> {
```

Comment:
It feels like this is gonna bite as back at some point, but I don't have a suggestion of how to simplify this.

Reply:
I actually think this will save us.

Reply:
Oh and by the way: if we implement lookups as operators in the expression language, most of this will go away.